File: ios/ExpoTappayDirectModule.swift
import ExpoModulesCore
import TPDirect

public class ExpoTappayDirectModule: Module {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
    
    
  private var tpdCard: TPDCard?
    
  public func definition() -> ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ExpoTappayDirect')` in JavaScript.
    Name("ExpoTappayDirect")

    Function("getTheme") { () -> String in
      "system"
    }
      
    Function("setup") { (appId: Int32, appKey: String, serverType: String) -> Void in
      let serverType: TPDServerType = (serverType == "production") ? .production : .sandBox
      TPDSetup.setWithAppId(appId, withAppKey: appKey, with: serverType)
    }
      
    Function("setCard") { (cardNumber: String, dueMonth: String, dueYear: String, ccv: String) -> Void in
      self.tpdCard = TPDCard.setWithCardNumber(cardNumber, withDueMonth: dueMonth, withDueYear: dueYear, withCCV: ccv)
    }
      
    Function("removeCard") { () -> Void in
      self.tpdCard = nil
    }
      
    AsyncFunction("getPrime") { (promise: Promise) in
      if let tpdCard = self.tpdCard {
        tpdCard
            .onSuccessCallback{ (prime, cardInfo, cardIdentifier, merchantReferenceInfo) in
              if
                  let directPayPrime = prime, directPayPrime != "",
                  let creditCardInfo = cardInfo,
                  let creditCardIdentifier = cardIdentifier
              {
                  promise.resolve([
                      "prime": directPayPrime,
                      "binCode": creditCardInfo.bincode ?? "",
                      "lastFour": creditCardInfo.lastFour ?? "",
                      "issuer": creditCardInfo.issuer ?? "",
                      "type": creditCardInfo.cardType,
                      "funding": creditCardInfo.funding,
                      "cardIdentifier": creditCardIdentifier
                  ])
              } else {
                  promise.reject("NO_PRIME", "PRIME IS EMPTY")
              }
            }
            .onFailureCallback{ (status, message) in
              promise.reject(String(status), message)
            }
            .createToken(withGeoLocation: "UNKNOWN")
      } else {
          promise.reject("NO_CARD", "PLEASE SET CARD FIRST")
      }
    }
  }
}

File: android/src/main/java/dev/stanma/tappay/direct/ExpoTappayDirectModule.kt
package dev.stanma.tappay.direct

import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import tech.cherri.tpdirect.api.TPDCard
import tech.cherri.tpdirect.api.TPDServerType
import tech.cherri.tpdirect.api.TPDSetup
import java.util.logging.Level
import java.util.logging.Logger

class ExpoTappayDirectModule : Module() {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.

  private var cardNumber: String = "";
  private var cardDueMonth: String = "";
  private var cardDueYear: String = "";
  private var cardCCV: String = "";

  override fun definition() = ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ExpoTappayDirect')` in JavaScript.
    Name("ExpoTappayDirect")

    Function("getTheme") {
      return@Function "system"
    }

    Function("setup") { appId: Int, appKey: String, env: String ->
      val serverType = when (env) {
        "sandbox" -> TPDServerType.Sandbox
        "production" -> TPDServerType.Production
        else -> TPDServerType.Sandbox
      }

      TPDSetup.initInstance(
        appContext.reactContext,
        appId,
        appKey,
        serverType,
      )
    }

    Function("setCard") { number: String, dueMonth: String, dueYear: String, ccv: String ->
      cardNumber = number
      cardDueMonth = dueMonth
      cardDueYear = dueYear
      cardCCV = ccv

      TPDCard(
        appContext.reactContext,
        StringBuffer(cardNumber),
        StringBuffer(cardDueMonth),
        StringBuffer(cardDueYear),
        StringBuffer(cardCCV),
      )
    }

    Function("removeCard") {
      cardNumber = ""
      cardDueMonth = ""
      cardDueYear = ""
      cardCCV = ""

      return@Function true
    }

    AsyncFunction("getPrime") { promise: Promise ->
      try {
        TPDCard(
          appContext.reactContext,
          StringBuffer(cardNumber),
          StringBuffer(cardDueMonth),
          StringBuffer(cardDueYear),
          StringBuffer(cardCCV),
        )
          .onSuccessCallback { prime, cardInfo, cardIdentifier, _ ->
            val primeData = mapOf(
              "prime" to prime,
              "binCode" to cardInfo.bincode,
              "lastFour" to cardInfo.lastFour,
              "issuer" to cardInfo.issuer,
              "type" to cardInfo.cardType,
              "funding" to cardInfo.funding,
              "cardIdentifier" to cardIdentifier
            )

            promise.resolve(primeData)
          }
          .onFailureCallback { status, reportMsg ->
            promise.reject(status.toString(), reportMsg, Exception(reportMsg))
          }
          .createToken("UNKNOWN")
      }catch (e: Exception) {
        promise.reject("ERROR", e.message, e)
      }
    }
  }
}
